### DISCUSSION TITLE
Adding support Unpack for routes args

### DISCUSSION THREAD

**Discussion by jag-k** [fastapi/fastapi#12519] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from pydantic import BaseModel, Field
from fastapi import FastAPI, Path, Query


app = FastAPI()


# The "old" format
# https://fastapi.tiangolo.com/tutorial/query-param-models/

class SearchRequest(BaseModel):
    """Search request"""

    title: str = Field(..., description="Search query")
    tags: list[str] = Field(default_factory=list, description="Task tags")



@app.get("/search")
async def search(search: Annotated[SearchRequest, Query()]) -> list:
    """Return search result."""
    return []


# The "new" format using TypedDict and python Unpack
# https://docs.python.org/3/library/typing.html#typing.Unpack

from typing import TypedDict, Unpack

class TypedSearchRequest(TypedDict):
    """Search request"""

    title: Annotated[str, Query(..., description="Search query")]
    tags: Annotated[list[str], Query(default_factory=list, description="Task tags")]



@app.get("/new-search")
async def new_search(**search: Unpack[TypedSearchRequest]) -> list:
    """Return search result."""
    return []
```

### Description

Recently, [Query Parameter Models](https://fastapi.tiangolo.com/tutorial/query-param-models/) were added to FastAPI, but we already have [`typing.Unpack` in Python](https://docs.python.org/3/library/typing.html#typing.Unpack).

You can find code examples of the old and my proposed version.

Personally, I think this can save us from additional checks and code duplication.

For example, this way you can add several main parameters at once (body, header, cookie, query, etc.), which are repeated in 2+ routes.

### Operating System

macOS

### FastAPI Version

0.115.3

### Pydantic Version

2.9.2

### Python Version

3.12.3

**Comment by shrey1605** [comment]
I am no expert but wouldn't you still need to define a class for each search request even in the case of using TypedDict + Unpack? If this is the case, then reduction in boilerplate code may not be that significant. Could you please elaborate on additional checks and code duplication part?

**Reply by jag-k** [reply]
Just for example from my real project:

```python
class TypedSearchRequest(TypedDict):
    """Search request"""

    title: Annotated[str, Query(..., description="Search query")]
    tags: Annotated[list[str], Query(default_factory=list, description="Task tags")]
    # Actually there are more parameters here
    response_mode: Annotated[Literal["base", "extended"], Header("base", description="Response mode")]



@app.get("/tasks")
async def tasks(current_user: UserRequiredDependency, **search: Unpack[TypedSearchRequest]) -> list:
    """Return tasks result."""


@app.get("/transcripts")
async def transcripts(admin: AdminRequiredDependency, **search: Unpack[TypedSearchRequest]) -> list:
   ...

@app.get("/organizations")
async def organizations(**search: Unpack[TypedSearchRequest]) -> list:
   ...

```

I hope this helps to understand my proposal.

**Reply by shrey1605** [reply]
Thank you for a more detailed example. I think it would be great if we can re-use the TypedDict like this across multiple search request params. I can see it being neater and potentially reducing the common code by quite a bit
